contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&key[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyEntryInitialize(&key[2], 0,
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(object_name));
+
+	scan = systable_beginscan(bbf_schema_rel,
+				get_bbf_schema_perms_idx_oid(),
+				true, NULL, 3, key);
+
+	tuple_bbf_schema = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple_bbf_schema))
+		catalog_entry_exists = true;
+
+	systable_endscan(scan);
+	table_close(bbf_schema_rel, AccessShareLock);
+	return catalog_entry_exists;
+}
+
+/*
+ * Get the value of permission column from BABELFISH_SCHEMA_PERMISSIONS table.
+ */
+int
+get_privilege_of_object(const char *schema_name,
+					const char *object_name,
+					const char *grantee,
+					const char *object_type)
+{
+	Relation	bbf_schema_rel;
+	HeapTuple	tuple_bbf_schema;
+	ScanKeyData	scanKey[5];
+	SysScanDesc	scan;
+	int16	dbid = get_cur_db_id();
+	int permission = 0;
+
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+									AccessShareLock);
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyEntryInitialize(&scanKey[2], 0,
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(object_name));
+	ScanKeyEntryInitialize(&scanKey[3], 0,
+				Anum_bbf_schema_perms_grantee,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(grantee));
+	ScanKeyEntryInitialize(&scanKey[4], 0,
+				Anum_bbf_schema_perms_object_type,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(object_type));
+	scan = systable_beginscan(bbf_schema_rel,
+				get_bbf_schema_perms_idx_oid(),
+				true, NULL, 5, scanKey);
+	tuple_bbf_schema = systable_getnext(scan);
+
+	if (HeapTupleIsValid(tuple_bbf_schema))
+	{
+		Form_bbf_schema_perms schemaform = (Form_bbf_schema_perms) GETSTRUCT(tuple_bbf_schema);
+		permission = schemaform->permission;
+	}
+
+	systable_endscan(scan);
+	table_close(bbf_schema_rel, AccessShareLock);
+	return permission;
+}
+
+/*
+ * Removes a row from the catalog BABELFISH_SCHEMA_PERMISSIONS.
+ */
+void
+remove_entry_from_bbf_schema_perms(const char *schema_name,
```

Comment:
Yes, it will be called at the time of DROP statement to remove all relevant entries.
We need to add some consistency check on this table. IMO let's handle it separately since it's already a lengthy PR.